pkg/apiserver/storage/vpc_inventory.go

```diff
@@ -0,0 +1,175 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+
+	antreastorage "antrea.io/antrea/pkg/apiserver/storage"
+	"antrea.io/antrea/pkg/apiserver/storage/ram"
+	"antrea.io/antrea/pkg/controller/types"
+	runtimev1alpha1 "antrea.io/nephe/apis/runtime/v1alpha1"
+)
+
+const (
+	VpcIndexerByAccountNameSpacedName = "namespace.accountname"
+	VpcIndexerByVpcNamespacedName     = "namespace.vpcname"
+	VpcIndexerByNamespace             = "vpc.namespace"
+	VpcIndexerByNamespacedRegion      = "namespace.region"
+
+	VpcLabelAccountName = "account-name"
+	VpcLabelRegion      = "region"
+)
+
+// vpcInventoryEvent implements storage.InternalEvent.
+type vpcInventoryEvent struct {
+	// The current version of the stored VPC.
+	CurrObject *runtimev1alpha1.Vpc
+	// The previous version of the stored VPC.
+	PrevObject *runtimev1alpha1.Vpc
+	// The key of this VPC.
+	Key             string
+	ResourceVersion uint64
+}
+
+// keyAndSpanSelectFunc returns whether the provided selectors matches the key and/or the nodeNames.
+func keyAndSpanSelectFunc(selectors *antreastorage.Selectors, key string, obj interface{}) bool {
+	// If Key is present in selectors, the provided key must match it.
+	if selectors.Key != "" && key != selectors.Key {
+		return false
+	}
+	// If nodeName is present in selectors's Field selector, the provided nodeNames must contain it.
+	if nodeName, found := selectors.Field.RequiresExactMatch("nodeName"); found {
+		if !obj.(types.Span).Has(nodeName) {
+			return false
+		}
+	}
+	return true
+}
+
+// isSelected determines if the previous and the current version of an object should be selected by the given selectors.
+func isSelected(key string, prevObj, currObj interface{}, selectors *antreastorage.Selectors, isInitEvent bool) (bool, bool) {
+	// We have filtered out init events that we are not interested in, so the current object must be selected.
+	if isInitEvent {
+		return false, true
+	}
+	prevObjSelected := !reflect.ValueOf(prevObj).IsNil() && keyAndSpanSelectFunc(selectors, key, prevObj)
+	currObjSelected := !reflect.ValueOf(currObj).IsNil() && keyAndSpanSelectFunc(selectors, key, currObj)
+	return prevObjSelected, currObjSelected
+}
+
+// ToWatchEvent converts the vpcEvent to *watch.Event based on the provided Selectors. It has the following features:
+// 1. Added event will be generated if the Selectors was not interested in the object but is now.
+// 2. Modified event will be generated if the Selectors was and is interested in the object but is now.
```

Comment:
Comment for modified and deleted event seem wrong, could you please rephrase them?Probably it should be
Modified event will be generated if the Selectors was and is interested in the object.
Deleted event will be generated if the Selectors was interested in the object but is not now.

Reply:
solved